File: 백준/Gold/23288. 주사위 굴리기 2/주사위 굴리기 2.swift
var input = readLine()!.split(separator: " ").compactMap{ Int(String($0))! }
var n = input[0]
var m = input[1]
var totalRoll = input[2]
var map = [[Int]]()

for _ in 0..<n{
    map.append(readLine()!.split(separator: " ").compactMap{Int(String($0))!})
    
}

var now = (0,0)
var dice = [4,1,3,2,5,6]

var directions = [[0,1],[1,0],[0,-1],[-1,0]]//동남서북
var di = 0

var roll = 0

var score = 0

while(roll<totalRoll){
    
    var nextPos = (now.0+directions[di][0],now.1+directions[di][1])
    if(0..<n ~= nextPos.0 && 0..<m ~= nextPos.1){
        di += 0
    }else{
        di=(di+2)%4 //방향반대로
        nextPos = (now.0+directions[di][0],now.1+directions[di][1])
    }
    
    now = nextPos
    
    switch(di){
    case 0:
        dice = rollToEast(dice: dice)
    case 1:
        dice = rollToSouth(dice: dice)
    case 2:
        dice = rollToWest(dice: dice)
    case 3:
        dice = rollToNorth(dice: dice)
    default:
        dice = rollToEast(dice: dice)
    }
    
    
    var below = dice[5]
    var boardNum = map[now.0][now.1]

    var queue = [(Int,Int)]()
    queue.append(now)
    var count = 1
    var visited = Array(repeating: Array(repeating: false, count: m), count: n)
    visited[now.0][now.1] = true
    while(queue.count>0){

        var nowNode = queue.removeFirst()
        
        for idx in 0..<4{
            var nr = nowNode.0 + directions[idx][0]
            var nc = nowNode.1 + directions[idx][1]
            if(0..<n ~= nr && 0..<m ~= nc && map[nr][nc] == boardNum && visited[nr][nc] == false){
                visited[nr][nc] = true
                count += 1
                queue.append((nr,nc))
            }else{
                continue
            }
        }
    }
    
    
    
    score += count*boardNum
    
    if(boardNum<below){
        di = (di+1)%4
    }else if(boardNum>below){
        di = (di+3)%4
    }else{
        di += 0
    }
    

    
    roll += 1
}


func rollToEast(dice: [Int])->[Int]{
    var new = [0,0,0,0,0,0]
    new[0] = dice[5]
    new[1] = dice[0]
    new[2] = dice[1]
    new[3] = dice[3]
    new[4] = dice[4]
    new[5] = dice[2]
    return new
}
func rollToWest(dice: [Int])->[Int]{
    var new = [0,0,0,0,0,0]
    new[0] = dice[1]
    new[1] = dice[2]
    new[2] = dice[5]
    new[3] = dice[3]
    new[4] = dice[4]
    new[5] = dice[0]
    return new
}
func rollToSouth(dice: [Int])->[Int]{
    var new = [0,0,0,0,0,0]
    new[0] = dice[0]
    new[1] = dice[3]
    new[2] = dice[2]
    new[3] = dice[5]
    new[4] = dice[1]
    new[5] = dice[4]
    return new
}
func rollToNorth(dice: [Int])->[Int]{
    var new = [0,0,0,0,0,0]
    new[0] = dice[0]
    new[1] = dice[4]
    new[2] = dice[2]
    new[3] = dice[1]
    new[4] = dice[5]
    new[5] = dice[3]
    return new
}

print(score)

File: 백준/Gold/23288. 주사위 굴리기 2/주사위 굴리기 2.py
def roll_to_east(dice):
    new_dice = [0, 0, 0, 0, 0, 0]
    new_dice[0] = dice[5]
    new_dice[1] = dice[0]
    new_dice[2] = dice[1]
    new_dice[3] = dice[3]
    new_dice[4] = dice[4]
    new_dice[5] = dice[2]
    return new_dice


def roll_to_west(dice):
    new_dice = [0, 0, 0, 0, 0, 0]
    new_dice[0] = dice[1]
    new_dice[1] = dice[2]
    new_dice[2] = dice[5]
    new_dice[3] = dice[3]
    new_dice[4] = dice[4]
    new_dice[5] = dice[0]
    return new_dice


def roll_to_south(dice):
    new_dice = [0, 0, 0, 0, 0, 0]
    new_dice[0] = dice[0]
    new_dice[1] = dice[3]
    new_dice[2] = dice[2]
    new_dice[3] = dice[5]
    new_dice[4] = dice[1]
    new_dice[5] = dice[4]
    return new_dice


def roll_to_north(dice):
    new_dice = [0, 0, 0, 0, 0, 0]
    new_dice[0] = dice[0]
    new_dice[1] = dice[4]
    new_dice[2] = dice[2]
    new_dice[3] = dice[1]
    new_dice[4] = dice[5]
    new_dice[5] = dice[3]
    return new_dice


input_values = input().split()
n, m, total_roll = map(int, input_values[0:3])
map_data = []

for _ in range(n):
    map_row = list(map(int, input().split()))
    map_data.append(map_row)

now = (0, 0)
dice = [4, 1, 3, 2, 5, 6]

directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]  # 동남서북
di = 0
roll = 0
score = 0

while roll < total_roll:
    next_pos = (now[0] + directions[di][0], now[1] + directions[di][1])

    if 0 <= next_pos[0] < n and 0 <= next_pos[1] < m:
        di += 0
    else:
        di = (di + 2) % 4  # 방향 반대로
        next_pos = (now[0] + directions[di][0], now[1] + directions[di][1])

    now = next_pos

    if di == 0:
        dice = roll_to_east(dice)
    elif di == 1:
        dice = roll_to_south(dice)
    elif di == 2:
        dice = roll_to_west(dice)
    elif di == 3:
        dice = roll_to_north(dice)

    below = dice[5]
    board_num = map_data[now[0]][now[1]]

    queue = [now]
    count = 1
    visited = [[False] * m for _ in range(n)]
    visited[now[0]][now[1]] = True

    while queue:
        now_node = queue.pop(0)

        for idx in range(4):
            nr = now_node[0] + directions[idx][0]
            nc = now_node[1] + directions[idx][1]

            if 0 <= nr < n and 0 <= nc < m and map_data[nr][nc] == board_num and not visited[nr][nc]:
                visited[nr][nc] = True
                count += 1
                queue.append((nr, nc))

    score += count * board_num

    if board_num < below:
        di = (di + 1) % 4
    elif board_num > below:
        di = (di + 3) % 4
    else:
        di += 0

    roll += 1

print(score)
